rln/src/iden3calc/graph.rs

```diff
@@ -0,0 +1,957 @@
+    let mut index = 0;
+    for (i, &used) in used.iter().enumerate() {
+        if used {
+            renumber[i] = Some(index);
+            index += 1;
+        }
+    }
+    assert_eq!(index, nodes.len());
+    for (&used, renumber) in used.iter().zip(renumber.iter()) {
+        assert_eq!(used, renumber.is_some());
+    }
+
+    // Renumber references.
+    for node in nodes.iter_mut() {
+        if let Node::Op(_, a, b) = node {
+            *a = renumber[*a].unwrap();
+            *b = renumber[*b].unwrap();
+        }
+        if let Node::UnoOp(_, a) = node {
+            *a = renumber[*a].unwrap();
+        }
+        if let Node::TresOp(_, a, b, c) = node {
+            *a = renumber[*a].unwrap();
+            *b = renumber[*b].unwrap();
+            *c = renumber[*c].unwrap();
+        }
+    }
+    for output in outputs.iter_mut() {
+        *output = renumber[*output].unwrap();
+    }
+
+    eprintln!("Removed {removed} unused nodes");
+}
+
+/// Randomly evaluate the graph
+fn random_eval(nodes: &mut Vec<Node>) -> Vec<U256> {
+    let mut rng = rand::thread_rng();
+    let mut values = Vec::with_capacity(nodes.len());
+    let mut inputs = HashMap::new();
+    let mut prfs = HashMap::new();
+    let mut prfs_uno = HashMap::new();
+    let mut prfs_tres = HashMap::new();
+    for node in nodes.iter() {
+        use Operation::*;
+        let value = match node {
+            // Constants evaluate to themselves
+            Node::Constant(c) => *c,
+
+            Node::MontConstant(_) => unimplemented!("should not be used"),
+
+            // Algebraic Ops are evaluated directly
+            // Since the field is large, by Swartz-Zippel if
+            // two values are the same then they are likely algebraically equal.
+            Node::Op(op @ (Add | Sub | Mul), a, b) => op.eval(values[*a], values[*b]),
+
+            // Input and non-algebraic ops are random functions
+            // TODO: https://github.com/recmo/uint/issues/95 and use .gen_range(..M)
+            Node::Input(i) => *inputs.entry(*i).or_insert_with(|| rng.gen::<U256>() % M),
+            Node::Op(op, a, b) => *prfs
+                .entry((*op, values[*a], values[*b]))
+                .or_insert_with(|| rng.gen::<U256>() % M),
+            Node::UnoOp(op, a) => *prfs_uno
+                .entry((*op, values[*a]))
+                .or_insert_with(|| rng.gen::<U256>() % M),
+            Node::TresOp(op, a, b, c) => *prfs_tres
+                .entry((*op, values[*a], values[*b], values[*c]))
+                .or_insert_with(|| rng.gen::<U256>() % M),
+        };
+        values.push(value);
+    }
+    values
+}
+
+/// Value numbering
+pub fn value_numbering(nodes: &mut Vec<Node>, outputs: &mut [usize]) {
+    assert_valid(nodes);
+
+    // Evaluate the graph in random field elements.
+    let values = random_eval(nodes);
+
+    // Find all nodes with the same value.
+    let mut value_map = HashMap::new();
+    for (i, &value) in values.iter().enumerate() {
```

Comment:
All of these for cycles below are not very idiomatic

Reply:
The cycle with HashMap does not need modifications, see the second part of https://stackoverflow.com/a/30441736 . The only loop in this function to be eliminated is the one at the line 596.